cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+        else:
+            self.indices = indices
+            # If indices is provided, it overrides self.n
+            self.n = len(indices)
+
+        self.L = L + int(preallocated)
+        self.shape = self.n, self.L, self.L
+
+        # Some client calls need to access the original filename(s) associated with a source
+        # These are traditionally available as the 'fname' attribute of the LazyImage class, hence only used by
+        # calling code when lazy=True
+        if filenames is None:
+            filenames = [""] * self.n
+        elif isinstance(filenames, str):
+            filenames = [filenames] * self.n
+        else:
+            assert len(filenames) == self.n
+            filenames = list(filenames)
+        self.filenames = np.array(filenames)
+
+        self.lazy = lazy
+        self.n_workers = n_workers
+        self.dtype = dtype
+
+        self.cache = None
+        if not self.lazy:
+            self.cache = ArraySource(self._images(np.arange(self.n)))
+
+    def __len__(self):
+        return self.n
+
+    def __getitem__(self, item):
+        return self.images(item)
+
+    def images(
+        self, indices: Optional[Union[int, np.ndarray, slice]] = None
+    ) -> torch.Tensor:
+        if indices is None:
+            indices = np.arange(self.n)
+        elif np.isscalar(indices):
+            indices = np.array([indices])
+        elif isinstance(indices, Iterable):
+            indices = np.array(np.fromiter(indices, int))
+        elif isinstance(indices, slice):
+            start, stop, step = indices.start, indices.stop, indices.step
+            start = start or 0
+            stop = min(stop, self.n) if stop is not None else self.n
+            step = step or 1
+            assert (
+                start >= 0 and stop >= 0 and step >= 0
+            ), "Only positive start/stop/step supported"
+            indices = np.arange(start, stop, step)
+        else:
+            raise TypeError("Unsupported Type for indices")
+
+        assert isinstance(indices, np.ndarray)
+
+        if np.any(indices >= self.n):
+            raise ValueError(f"indices should be < {self.n}")
+
+        # Convert incoming caller indices to indices that this ImageSource will use
+        indices = np.array(self.indices[indices])
+
+        if indices.size == 1:
+            logger.debug(f"ImageSource returning images for index {indices}")
+        else:
+            logger.debug(
+                f"ImageSource returning images for {len(indices)} indices ({indices[0]}..{indices[-1]})"
+            )
+
+        if self.cache:
+            images = self.cache._images(indices)
+        else:
+            images = self._images(indices).astype(self.dtype)
+        return torch.tensor(images)
+
+    def _images(self, indices: np.ndarray) -> np.ndarray:
+        raise NotImplementedError("Subclasses must implement this")
+
+    def chunks(self, chunksize=1000):
+        return ImageSourceIterator(self, chunksize=chunksize)
```

Comment:
True - very nice. A slightly different form works in our case. Added in b2f85194f01817c24ca579a66f6014628930d29e.
```
        for chunk in torch.split(torch.arange(self.n), chunksize):
            yield chunk, self.images(chunk)
 ```

Reply:
Ah - I forgot `indices` are ndarrays, not Tensors. `pyright` caught this. `np.array_split` works differently, so it gets slightly annoying, but an equivalent approach to `torch.split` is implemented in commit 13bf8bcbc5881d191bb7ac687e0449177f919a55
```
        split_indices = np.arange(chunksize, self.n, chunksize)
        for chunk in np.array_split(np.arange(self.n), split_indices):
            yield chunk, self.images(chunk)
```

If we don't care if the returned chunks may be off-by-one in size (which may be the case for the last chunk anyway), then we can use `self.n // chunksize` and the implementation is simpler.

Also note that `self.indices` is out of the picture here - that's something internal to `images()`.

Reply:
"not to make unnecessary copies of the data"

This sounds like premature optimization to me. A copy is essentially free relative to the cost of running a forward pass on the data. And views are somewhat dangerous (remember what happened with the in-place division!)

Reply:
Point taken. I've replaced the recent tweak to `images()` behavior such that any invocation of `images()` will return a copy of the data, whether the `ImageSource` is lazy or not (in the latter case it will be much faster of course). Implemented in 194f845863403eac824355ad33b11f6eff737946